### PR TITLE
Unescape strings when parsing desktop entries

### PR DESCRIPTION
**Pull request by lePerdu**
This is based of the [spec](https://standards.freedesktop.org/desktop-entry-spec/latest/ar01s03.html) and works with all of my desktop files (including those created by wine).

### CODE REVIEW

File: lib/menubar/utils.lua [hunk 1]
```diff
@@ -151,6 +168,68 @@ function utils.rtrim(s)
...
+-- the usual meanings.
+-- @tparam string s String to unescape
+-- @tparam boolean list Whether the string should be escaped as a list
+-- @treturn string|table The unescaped string or a table of unescaped strings if
+-- list is true.
+function utils.unescape(s, list)
+    if not s then return end
+    -- We can't just use string.gsub because something like [[\\s]] would become
+    -- either [[ ]] or [[\ ]] (depending on the order) while the correct result
+    -- is [[\s]].
```
Comment (psychon) [inline lib/menubar/utils.lua]: How about (basically untested):
```
return string.gsub(s, "\\.", function(sequence)
    local c = sequence:sub(2, 2)
    return escape_chars[sequence:sub(2, 2)] or sequence
end)
```
The list case is a bit harder (should it get its own, separate function? this does more than just handling escape sequences). Again, some untested idea:
```
local segments = {}
local part = ""
-- Iterate over sub-strings between semicolons
for word in string.gmatch(s, "[^;]+") do
  if word:sub(-1, -1) == "\\" then
    -- The semicolon was escaped, remember this part for later
    part = part .. word:sub(1, -2) .. ";"
  else
    table.insert(segments, utils.unescape(part .. word))
    part = ""
  end
end
```

File: tests/test-menubar.lua [hunk 2]
```diff
@@ -55,6 +55,56 @@ runner.run_steps {
...
+                if returned[i] ~= unescaped[i] then
+                    return false
+                end
+            end
+        end
+
+        return true
+    end,
+
     function()
```
Comment (psychon) [inline tests/test-menubar.lua]: I'd prefer this as unit tests instead of integration tests, i.e. in `spec/menubar/utils_spec.lua`.

Comment (lePerdu) [reply tests/test-menubar.lua]: Yeah, it fits better there. Didn't know about that directory.

File: lib/menubar/utils.lua [hunk 3]
```diff
@@ -169,68 +169,48 @@ function utils.rtrim(s)
...
+    if not string.match(s, ';%s*$') then
+        s = s .. ';'
+    end
 
-        return s
+    local segments = {}
+    local part = ""
+    -- Iterate over sub-strings between semicolons
+    for word, backslashes in string.gmatch(s, "([^;]-(\\*));") do
+        if #backslashes % 2 ~= 0 then
```
Comment (psychon) [inline lib/menubar/utils.lua]: Oh right, I missed that part. Nice catch & nice fix.

File: lib/menubar/utils.lua [hunk 4]
```diff
@@ -169,68 +169,48 @@ function utils.rtrim(s)
...
+-- Semi-colons in lists are escaped as '\;'.
+-- @tparam string s String to parse
+-- @treturn table A table containing the separated strings. Each element is
+-- unescaped by utils.unescape().
+function utils.parse_list(s)
+    if not s then return end
+
+    -- Append terminating semi-colon if not already there. Ignore whitespace
+    -- at the end
+    if not string.match(s, ';%s*$') then
```
Comment (psychon) [inline lib/menubar/utils.lua]: Wouldn't `if string.sub(s, -1, -1) ~= ';'` be better? The spec does not seem to talk about ignoring trailing whitespace (so I guess we should have a test for that case).

Comment (lePerdu) [reply lib/menubar/utils.lua]: I took "trailing empty strings" to mean whitespace also, but I guess that's not technically correct and `desktop-file-utils` from freedesktop.org does not ignore trailing whitespace. I'll change that and add the test.

**Comment by blueyed** [comment]
Thank you!

**Pull request merged**